### ISSUE TITLE
python-mypy Suspicious state from syntax checker when navigating to site-packages code

### ISSUE THREAD

**Issue by dsyzling** [flycheck/flycheck#1705]
## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].
- [x] I have searched existing issues and StackOverflow for solutions.

## Bug description
flycheck 20200320.1402
mypy 0.770
python 3.7.6
Emacs 26.2 on Arch Linux

Navigating to a python file that may be within site-packages via 'goto definition' (lsp or xref) causes python-mypy to report:

Suspicious state from syntax checker python-mypy: Flycheck checker python-mypy
returned non-zero exit code 1, but its output contained no errors:
/home/dsyzling/miniconda3/envs/pandastest/lib/python3.7/site-packages is in the
PYTHONPATH. Please change directory so it is not.

Now this appears to be caused due to the full path of the file being passed to mypy and mypy detects site-packages in the path, rather than checking the file it claims site-packages is in the PYTHONPATH - which it isn't. If you were to navigate to the directory and run mypy on the file without the full path it type checks the file and shows errors. 

So mypy complains and flycheck sees the return code of 1 without any individual checker errors so outputs the message above. Here's an example run with C-c ! C-c:

mypy --show-column-numbers /home/dsyzling/miniconda3/envs/pandastest/lib/python3.7/site-packages/pandas/core/frame.py
/home/dsyzling/miniconda3/envs/pandastest/lib/python3.7/site-packages is in the
PYTHONPATH. Please change directory so it is not.

The above can occur when I go to the definition of DataFrame for example in pandas to review the source or check the docs. 

Now it should be said I don't necessarily want flycheck to run on these library files, I'm browsing the source code of libraries. Currently however navigating to that directory and leaving the buffer open, the warning message is constantly displayed from flycheck. Closing the buffer clears the problem. 

In some ways I'd like to prevent type checking of the source file and reporting errors to me, but I don't want to tell mypy to ignore type checking from the point of view of importing types from a third party library so that when I use functions/classes from the library I will be warned of type mismatches. 

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

**Comment by dsyzling** [comment]
I raised the issue as a bug but I'm not entirely convinced it's a bug with flycheck, it's unfortunate as the result is a persistent nagging message that won't clear until that buffer is closed.

I'm also not sure of the behaviour I actually want - except possibly in this case I don't necessarily need to syntax/lint check these files. But then how do we exclude/ignore them, should the option be in mypy, lsp-mode, flycheck?

I will raise a question over at mypy and see what they say. You're right there's a certain inconsistency there and it may be a side effect that they are unaware of.

When people use lsp with Palentir (pyls), Mypy is invoked as a plugin through pyls and the results sent back via lsp diagnostic messages. I suspect they just ignore the fact that there's no errors been reported even though the command returned a non zero return code.  The issue cropped up for me because I was experimenting with the Microsoft Python language server (which doesn't support linters as plugins) but still wanted flake8, mypy checking via flycheck. This isn't really supported by lsp-mode at the moment since lsp wants to be the flycheck checker so the remaining checkers aren't necessarily invoked.

**Comment by cpitclaudel** [comment]
The easy way out for us would be if mypy just accepted to run on those files.  Otherwise, if mypy gives us a way to determine that it won't run on those files, we can use that to deactivate it.  Finally, if mypy used its standard error output format, we would just report that error.

Of course, we can implement a workaround; in this case, it would be to add an error pattern for this specific message.

**Comment by cpitclaudel** [comment]
Hi @dsyzling , did you get an opportunity to report the issue to mypy?

**Comment by dsyzling** [comment]
Yes sorry - I've reported the issue, they've agreed there's a problem there although they're not sure quite where. An initial trace run bailed early and didn't report anything useful so it will require a little more digging. 

Since it's lower down the list of priorities I said I would try and take a look and try and gather some further debug info from mypy and/or see if I can look at options for a fix. Just trying to find some time to take a look.

**Comment by dsyzling** [comment]
I've submitted further debug information to the mypy team. One of the developers believes this issue will be resolved with the merge of issue 5701 here - https://github.com/python/mypy/issues/5701. So I've agreed to retest with flycheck when the code is merged into mainline or is available for testing. 

Hopefully this will mean there's nothing to do on the flycheck side. So I'm happy if you want to close this issue, I'll monitor mypy and work with them for a resolution - even if that ends up being writing the error message to stderr, but hopefully we'll have something more useful.

**Comment by cpitclaudel** [comment]
Thanks a lot for the update; it's much appreciated.
One thing that worries me a bit from reading that thread is the comment about the current directory (the one where Flycheck is invoked) being added to PYTHONPATH.  How are we to guess where mypy is supposed to be invoked from? We've discussed this in https://github.com/flycheck/flycheck/issues/1711, but I fear our fix isn't complete.

**Comment by dsyzling** [comment]
I'm not sure either, since the work is still to be done it may or may not resolve this issue and could possibly throw up new problems related to #1711. I can only really wait for their updated implementation, retest and then possibly reengage to discuss the wider issue.

**Comment by bbatsov** [comment]
Closing as upstream. Mypy intentionally rejects files located in site-packages, so the "suspicious state" reported here is expected upstream behavior. There is nothing Flycheck can do to change how mypy handles these paths. Thank you for the report.
